### DISCUSSION TITLE
Struggling with layout of TestStampStyle

### DISCUSSION THREAD

**Discussion by ghost** [MatthiasValvekens/pyHanko#298] [General]
Hi,
I'm struggling with the layout of my visual signatures or more specifically with the TextStampStyle object.

This is my code for signing a document. The signature box comes from a parameter (annotatedRect) or from an existing signature field selected by the parameter signatureFieldName:
```
def signDocumentInPlace(documentPath, signatureFieldName, annotatedRect, annotatedPage, pfxPath, pfxPin): 
    cms_signer = signers.SimpleSigner.load_pkcs12(
		pfx_file = pfxPath, 
		passphrase = pfxPin  
	)
    if cms_signer == None:
        LOGGER.error("Error while opening PFX file.")
        raise PFXException()

    with open(documentPath, 'rb+') as doc:
        out = None
        fields = None
        fieldName = None
        pdfFile = IncrementalPdfFileWriter(doc, strict = False)
        signature_meta = None
        if signatureFieldName != None:
            fieldName = signatureFieldName
        else:
            rnd = random.randrange(1000, 9999)
            fieldName = "Sig" + str(rnd)
            append_signature_field(
                pdfFile, 
                SigFieldSpec(
                    sig_field_name = fieldName, 
                    on_page = annotatedPage - 1, 
                    box = (annotatedRect['x0'], annotatedRect['y0'], annotatedRect['x1'], annotatedRect['y1']),
                ),
            )
            

        signature_meta = signers.PdfSignatureMetadata(
            field_name = fieldName,
            md_algorithm = 'sha256',
            subfilter = SigSeedSubFilter.PADES, #ADOBE_PKCS7_DETACHED, 
            validation_context = getValidationContext()[0],
            signer_key_usage={'digital_signature'},
            embed_validation_info = True,
            use_pades_lta = False # PAdES signature level: False -> B-LT, True -> B-LTA
        )
				
        
        pdfSigner = signers.PdfSigner(
            signature_meta = signature_meta, 
            signer = cms_signer,
            timestamper = timestamps.HTTPTimeStamper(TIMESTAMP_SERVER_URL),
            stamp_style = TextStampStyle(
                stamp_text = 'Digitally signed by\n' + cms_signer.subject_name.split('<')[0] + '\non %(ts)s',
                timestamp_format = 'Date: %Y-%m-%d %H:%M:%S %Z',
                text_box_style = TextBoxStyle(
                    font = GlyphAccumulatorFactory(os.path.join(STATIC_ROOT, "manager/fonts/NotoSans-Regular.ttf")),
                    box_layout_rule = SimpleBoxLayoutRule(x_align = AxisAlignment.ALIGN_MID, y_align = AxisAlignment.ALIGN_MID, inner_content_scaling=1, margins = Margins(2,2,2,2)), 
                ),
                inner_content_layout = SimpleBoxLayoutRule(x_align = AxisAlignment.ALIGN_MIN, y_align = AxisAlignment.ALIGN_MID, inner_content_scaling = 3),
                background = PdfImage(os.path.join(STATIC_ROOT, "manager/images/Logo.png"), pdfFile),
                background_opacity = 0.2,						
                border_width = 1,
            ),
        )
        out = pdfSigner.sign_pdf(
            pdfFile,
            in_place = False,
            existing_fields_only = False,
        )
        with open(documentPath, 'wb') as out_doc:
            out_doc.write(out.getbuffer())
```

What I want is a signature that:
- Shows a logo (Logo.png) in the background
- Shows a text with several line breaks on top of the logo
- The logo shall be centered in the rect and scaled as large as possible while keeping the ratio between the width and height of the PNG file
- The text block shall be centered and scaled as large as possible in the rect while keeping the aspect ratio of the font. So scaling should happen via font size. 

My problems are:
- With the code above the font of the displayed text looks squeezed or stretched depending on the size of the rect. It seems that the text is first rendered and the resulting bitmap is then squeezed into the rect. Width and height are not necessarily scaled with the same factor for the fitting. 
- Another independent problem: When I use a standard font like Arial, there seems to be too much space between the characters and lines are not aligned after line breaks. Sometimes they are cut off at the rect borders.

Can somebody give me a tip how to solve this problem?

Thanks Tuelle

**Comment by MatthiasValvekens** [comment]
Hi @tuelle,

Yes, the layout code was cobbled together a bit too naively and has quite a few sharp edges that are in need of addressing. I should probably add that to the known issues page...

In particular, the font size parameter is very misleading, since pyHanko's layouter (as implemented today) will simply try to fit the text in the box and (effectively) only uses the font size to do OpenType magic stuff that happens to depend on the font size for glyph rendering specifics. I agree that this state of affairs is very confusing and not helpful at all. I would welcome PRs that address this class of problems more directly, but that task is probably not for the faint of heart...

Anyway, I can't really run any tests right where I am right now, so I can't give you any concrete pointers for the time being.

To your second point:

> When I use a standard font like Arial, there seems to be too much space between the characters and lines are not aligned after line breaks. Sometimes they are cut off at the rect borders.

This is (semi-)expected. With non-embedded fonts, you kinda still need access to (approximately) correct font metrics to produce accurate output. PyHanko doesn't ship with font metric files even for the standard fonts, so results will be poor unless you make the actual font file available to pyHanko and let it be embedded. Again, I wouldn't object to PRs rectifying this---it's probably not particularly invasive to add, but it requires moderate implementation effort and quite a bit of knowledge of the way fonts are represented in PDF.

FWIW, this is the reason why I set the default font to a fixed-pitch one... 🙃

**Comment by ghost** [comment]
Hi Matthias,
Understood! There are probably more interesting topics to spend time on. 

I have started render a signature graphic (text with background image) with Pillow. The plan is to set the resulting signature graphic as a background image in a TextStamp and leave the text of the TextStamp empty. I did not check yet if this is possible. 

I'm nevertheless wondering how to make sure that the timestamp in the graphic is exactly the same as the one in the cryptographic information of the signature. If I understand your code correctly you're also using now() to generate the time stamp in the signature text. So there is the risk chance that time in the signature graphic and in the time cryptographic information differ by a second, right? This seems to be a chicken-egg problem because the signature graphic must be available when you generate the actual signature, mustn't it. Or is there a trick?

Thanks for your support!
Tuelle

**Reply by MatthiasValvekens** [reply]
I'll make a note to look into the layout code more deeply once I'm able to (currently travelling). If nothing else, that should allow me to improve the _documentation_ at the very least. 😅 

In the meantime: if you're very keen on replicating an exact layout, note that pyHanko allows you to supply an externally-generated PDF file from which you can extract a page to use as the signature appearance. So you could generate the appearance using `cairo` (or some other 2D graphics toolkit) and supply it to pyHanko as a pregenerated asset. [There's some documentation over here](https://pyhanko.readthedocs.io/en/latest/lib-guide/signing.html#static-content-stamps). The asset will be imported as-is from the source PDF, so if the source has text/vectors/embedded fonts/... you'll get all of that in the resulting signed file as well.

(Actually, just thinking out loud: maybe setting up a proper `PdfContent` <> `pycairo` integration could get us out of this mess as well... I'll chew on that for a bit.)


> I'm nevertheless wondering how to make sure that the timestamp in the graphic is exactly the same as the one in the cryptographic information of the signature. If I understand your code correctly you're also using now() to generate the time stamp in the signature text. So there is the risk chance that time in the signature graphic and in the time cryptographic information differ by a second, right? This seems to be a chicken-egg problem because the signature graphic must be available when you generate the actual signature, mustn't it. Or is there a trick?

This is a good question, and you're right about the chicken-and-egg problem here. However, in practice it's not a huge deal: as long as you generate the CMS signature container locally, you're in control of the clock, so there's effectively nothing stopping you from forcing both timestamps to be equal. Checking quickly, it looks like pyHanko supports this, but not at the highest level of the API, and it's poorly documented/untested: you can override `system_time` on `PdfSigningSession`. I could streamline that a bit if it's important to you.

**EDIT: actually, looking more closely, it seems that both the appearance generation code and the CMS generation code already read from this `system_time` attribute.** So they should be equal already if you use "default everything". If not, that's definitely a bug, and I'd be very keen to know how to reproduce it :).

----------------

Side note: the fact that this time manipulation is possible also illustrates why it's not really a big deal security-wise either: both timestamps are untrusted information from the PoV of the validator anyway, since they're "naked" assertions based on a clock under the (presumptive!) signer's control. This has two implications:

- In some cases, the "PDF signature assembly" process (=pyHanko's core task) is actually _not_ under the operational control of the same entity as the one that puts together the signature container. This comes up with some frequency in remote signing implementations. This difference can cause deviations in the claimed signature time ranging from a few seconds to even multiple minutes, e.g. if the remote signer triggers some kind of manual validation process that involves a human having to acknowledge something on their phone.
- When it matters (i.e. in all cases where the signature needs to outlive the signer's certificate), you'd include an additional signature timestamp token from a publicly trusted time stamping authority. And then you get the same chicken-or-egg problem: the timestamp token (by definition) has to be requested _after_ the signature has been generated, so it will also be off by some (usually small) delta.

TL;DR: no validator is going to care about this kind of discrepancy in practice, and it's not always avoidable.
